services/oauth/v2/oauth.go

```diff
@@ -0,0 +1,396 @@
+- If the token is not present in the cache, it fetches the token from the control plane
+- It returns the status code, token and error
+- It also sends the stats to the statsd
+- It also sends the error to the control plane
+Parameters:
+  - fetchTokenParams: *RefreshTokenParams
+  - logTypeName: string
+  - authStats: *OAuthStats
+
+Return:
+  - int: status code
+  - *AuthResponse: token
+  - error: error
+*/
+func (h *OAuthHandler) FetchToken(fetchTokenParams *RefreshTokenParams) (int, *AuthResponse, error) {
+	authStats := &OAuthStats{
+		id:              fetchTokenParams.AccountId,
+		workspaceId:     fetchTokenParams.WorkspaceId,
+		rudderCategory:  "destination",
+		statName:        "",
+		isCallToCpApi:   false,
+		authErrCategory: "",
+		errorMessage:    "",
+		destDefName:     fetchTokenParams.DestDefName,
+		isTokenFetch:    true,
+		flowType:        h.RudderFlowType,
+		action:          "fetch_token",
+	}
+	return h.GetTokenInfo(fetchTokenParams, "Fetch token", authStats)
+}
+
+/*
+Refresh token function is used to refresh the token from the control plane
+- It fetches the token from the cache
+- If the token is present in the cache, it checks if the token has expired
+- If the token has expired, it fetches the token from the control plane
+- If the token is present in the cache, it compares the token with the received token
+  - If it doesn't match then it is possible the stored token is not expired but the received token is expired so return the stored token
+  - Else if matches then go ahead
+
+- It fetches the token from the control plane
+- It returns the status code, token and error
+- It also sends the stats to the statsd
+- It also sends the error to the control plane
+Parameters:
+  - refTokenParams: *RefreshTokenParams
+
+Return:
+  - int: status code
+  - *AuthResponse: token
+  - error: error
+*/
+func (h *OAuthHandler) RefreshToken(refTokenParams *RefreshTokenParams) (int, *AuthResponse, error) {
+	authStats := &OAuthStats{
+		id:              refTokenParams.AccountId,
+		workspaceId:     refTokenParams.WorkspaceId,
+		rudderCategory:  "destination",
+		statName:        "",
+		isCallToCpApi:   false,
+		authErrCategory: CategoryRefreshToken,
+		errorMessage:    "",
+		destDefName:     refTokenParams.DestDefName,
+		flowType:        h.RudderFlowType,
+		action:          "refresh_token",
+	}
+	return h.GetTokenInfo(refTokenParams, "Refresh token", authStats)
+}
+
+func (h *OAuthHandler) GetTokenInfo(refTokenParams *RefreshTokenParams, logTypeName string, authStats *OAuthStats) (int, *AuthResponse, error) {
+	h.Logger.Debugn("[request] :: Get Token Info request received",
+		logger.NewStringField("Module name:", h.LoggerName),
+		logger.NewStringField("Call Type", logTypeName),
+		logger.NewStringField("AccountId", refTokenParams.AccountId),
+		obskit.DestinationID(refTokenParams.Destination.DestinationId),
+		obskit.WorkspaceID(refTokenParams.WorkspaceId),
+		obskit.DestinationType(refTokenParams.DestDefName))
```

Comment:
We can create a parameterized logger once and use it?
```suggestion
        log := h.Logger.Withn(
		logger.NewStringField("Module name:", h.LoggerName),
		logger.NewStringField("Call Type", logTypeName),
		logger.NewStringField("AccountId", refTokenParams.AccountId),
		obskit.DestinationID(refTokenParams.Destination.ID),
		obskit.WorkspaceID(refTokenParams.WorkspaceId),
		obskit.DestinationType(refTokenParams.DestDefName),
	)
	
	log.Debugn("[request] :: Get Token Info request received")
	log.Debugn("[request] :: Failed to type assert the stored cache")
```